[Ready to Review] Quick search projects

pattisdr opened the pull request:
# Purpose

Main Issue: https://openscience.atlassian.net/browse/OSF-5382

Covers both subtasks:
https://openscience.atlassian.net/browse/OSF-5383
https://openscience.atlassian.net/browse/OSF-5384

Displays user's projects (10 by default) with option to load more.  User can search and sort projects.  Logs/comments since last logged in functionality has been postponed until we start properly tracking this information.
# Changes

New dashboard change.
- First ten projects load and then view is rendered.  Remaining projects are recursively loaded in the background.
- Until all the projects are loaded, user cannot search or sort through them.  
- Default projects sort is date_modified, most recent first
- Can sort by modified/title ascending/descending
- If >3 contributors, formatting is: Pattison, Pattison, and Pattison + 23
- Can search on title, contributor full name, and tags
- Clear search button
- Search results display as you type
- Clear search/backspace restores results prior to search.
- Responsive.  As screen narrows, title column widens.  For xs screens, nodes stack.
## My implementation

![screen shot 2016-01-15 at 10 23 50 am](https://cloud.githubusercontent.com/assets/9755598/12356738/1eaa7e6c-bb72-11e5-94d1-7d445a397d44.png)
## Wireframe

![screen shot 2016-01-15 at 11 05 56 am](https://cloud.githubusercontent.com/assets/9755598/12357821/1581e270-bb78-11e5-868b-aaa82d5a3f97.png)

(example above, no it's not the same 'Alphabet' project showing up, these Alphabet nodes were bulk created at the same time)

pattisdr committed:
Rename home-page to landing_page. Add home-page.js. Update webpack.common.config.js.

Add quickProjectSearchPlugin.

pattisdr committed:
Add script to mako template.

pattisdr committed:
Mount quickSearchProjectPlugin to home-page.

pattisdr committed:
Add addQuickProjectSearchWrap element for getElementById.

pattisdr committed:
Have mithril view return node titles from API call users/me/nodes/

pattisdr committed:
Add contributors to project entries

pattisdr committed:
Consolidate getContributors method.

pattisdr committed:
Change loading of nodes so it is not a function.

pattisdr committed:
Add functionality to initially load ten projects and to load ten more on button click.

pattisdr committed:
Format date on date_modified

pattisdr committed:
Display first ten nodes automatically, and then add up to ten at a time when button clicked.

pattisdr committed:
Remove 'Load More' button once all nodes loaded.

pattisdr committed:
Add sort ascending, descending on title, date modified.

pattisdr committed:
Original nodes auto-sort on date_modified.

pattisdr committed:
Add capability to retrieve last_login date.  Add field to serializer.

pattisdr committed:
Add capability to retrieve comments since last login date

pattisdr committed:
Add search bar placeholder, make title link to project.

pattisdr committed:
Add functionality for getting log counts

pattisdr committed:
Properly add filtering on date_modified and comment_added to params.

pattisdr committed:
Temporary lastLogin for testing.

pattisdr committed:
Temp last_login

pattisdr committed:
Add glyphicon chevrons for sort buttons.

pattisdr committed:
numContributors not displayed is a number.

pattisdr committed:
Bootstrap grid for easier viewing - temporary.

pattisdr committed:
Add countState and sortState parameters.

pattisdr committed:
Add param to restoreToNodeList, so the nodes sent in as param will be restored to the node list.

pattisdr committed:
Add method to quick search projects on title.

pattisdr committed:
If search box is cleared, restore nodes in view to what was visible prior to search.  Add method to sort nodes  by previous field.

pattisdr committed:
When sort button pressed, will sort all nodes excluding nodes that don't match search(if any).  When returning search results, use same sorting mechanism that was selected prior to search.

pattisdr committed:
Variable explanations

pattisdr committed:
Extend page size for now.

pattisdr committed:
Formatting.

pattisdr committed:
Organization.

pattisdr committed:
Add filtering on contributor names, even contributor names that are not visible.   Rearrange functions for filtering on title and contributor.

pattisdr committed:
Add formatting for three contributors.

pattisdr committed:
Add button to show less nodes
Decrements by 10 if both divisible by ten and greater than 10.
Example: You have 24 projects.  All projects are loaded.  Clicking 'show less' will remove 4.  Now 20 are shown.  Clicking 'show less' will remove 10, now.
If less than 10 nodes displayed, show less button disappears.
Basically, it undoes what the show more button did.

pattisdr committed:
Turn node list into a m.prop

pattisdr committed:
Make displayedNodes, nonMatchingNodes, countState, and sortState m.prop()

pattisdr committed:
Change how nodes are loading. Pull in first ten nodes, render view, then load remaining nodes recursively.

pattisdr committed:
Hide buttons for sorting, searching until all nodes loaded.

pattisdr committed:
API change so users/me/nodes/ will sort by -date_modified.

pattisdr committed:
Default sort state of nodes is dateDesc.

pattisdr committed:
Set state as allLoaded and redraw when there are less than ten projects.

pattisdr committed:
Retrieve contributors as retrieving nodes and place in contributorMapping {node_id: [contributor names]} for later searching.

Prior to this, nodes were being called with a page[size]=10000, which cascaded down to embedded contributors.  This insured embedded contributors were present.  Now I need to make sure all contributors are retrieved.

pattisdr committed:
If user has no nodes (and there's no filter on the nodes), return No nodes found.

pattisdr committed:
Display helpful message if sort returns no results.

pattisdr committed:
Remove functionality for logs since last logged in, and comments since last logged in.

We are not tracking date_last_login currently, only date since last request.

pattisdr committed:
Remove bootstrap formatting.

pattisdr committed:
Original countState should be result of first API call to users/me/nodes.

pattisdr committed:
Add spinner until all projects loaded.

pattisdr committed:
Remove last_login field from UserSerializer.

pattisdr committed:
Redraw view when more nodes are loaded so the show more button is available.  Only want the show more button visible if there are nodes that can be loaded. Hide the `show less` button until all nodes are loaded.

pattisdr committed:
Make comment more clear.

pattisdr committed:
Remove earlier formatting of QuickSearchProject container

pattisdr committed:
Turn resultsFound results into divs instead of table.

pattisdr committed:
Turn modular nodes into divs instead of table rows.

pattisdr committed:
Add magnifying glass icon to search bar.

pattisdr committed:
Add outline to node representation

pattisdr committed:
Turn 'load more' and 'load less' buttons into carets.

pattisdr committed:
For sort buttons, use font awesome icons instead of glyph icons.

pattisdr committed:
Increase size of 'load_more' and 'load_less' buttons.

pattisdr committed:
Change column headers to bold.  Move "My Projects" title to plugin.

pattisdr committed:
Add button styling so only font-awesome icon visible, not button outline.

pattisdr committed:
Remove "My Projects" h1 header from mako template.

pattisdr committed:
move projectView closer to resultsFound.

pattisdr committed:
Move node representation outline to exclude outer two columns.

pattisdr committed:
Formatting when no results found.

pattisdr committed:
Formatting when you have no projects.

pattisdr committed:
Use angle-up, angle-down instead of chevron-up, chevron-down.

pattisdr committed:
Add very light gray to node background with a med gray border.

pattisdr committed:
Remove container from outer div.

pattisdr committed:
Change formatting of nodes so outer border is on row instead of individual columns.  Use an offset to narrow the row display.

pattisdr committed:
Change formatting to small device.

pattisdr committed:
Have sortDateDesc button automatically highlighted.  Add functionality to highlight other sorting buttons as you click them and deselect the others.

pattisdr committed:
Shrink load nodes icons.

pattisdr committed:
Add hovering to nodes.

pattisdr committed:
If no projects found, still need to have "My Projects" heading.

pattisdr committed:
Remove formatting from search box, so loading spinner and search box take up same footprint.  Use OSF spinner.

pattisdr committed:
Change column spacing to give date modified more room.

pattisdr committed:
Remove 'load less' button.  Determined that users are not into restorative actions.

pattisdr committed:
Rename node 'name' to 'title'.  User could confuse "name" with contributor name.

pattisdr committed:
Add filtering on tags in addition to contributors and title.

pattisdr committed:
Make entire node div a link to the node instead of having just the title link to the node.

pattisdr committed:
Implement a clear search button.

pattisdr committed:
When shrink window to extra small, have column headers disappear.

pattisdr committed:
When shrink window, have sorting become a dropdown for the field(Title, Modified), and arrows for Ascending, Descending.  Dropdown default value is already set to whatever the field sort was before shrinking or 'date_modified'.  Default Ascending/Descending arrow also highlighted.

pattisdr committed:
Remove bold from label.

pattisdr committed:
As shrink screen, give more space to title.

pattisdr committed:
Add comments to js functions.

pattisdr committed:
Merge branch 'dashboard-redesign-component-3' of https://github.com/caneruguz/osf.io into quickSearchProjects

Conflicts:
	api/users/views.py

pattisdr committed:
Use bootstrap form-control for sort dropdown.

pattisdr committed:
For users/me/nodes change query to have is_collection is not equal to true instead of is_folder.  Was overwritten in merge.

pattisdr committed:
Take registrations out of users/me/nodes.

pattisdr committed:
Remove folders from users/me/nodes.

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+        self.nodes = m.prop([]); // Pending nodes waiting to be displayed
+        self.displayedNodes = m.prop([]); // Nodes that are rendered
+        self.nonMatchingNodes = m.prop([]); //Nodes that don't match search query
+        self.sortState = m.prop('dateDesc');
```
The comments after these helps a lot. Can you add comments at the end of all of them?

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+        promise.then(
+            function(){
+                if (self.next()) {
+                    self.recursiveNodes(self.next())
```
Can you add an initial value to the top with a description comment for self.recursiveNodes()?

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+                    else {
+                        self.loadingComplete(true);
+                        m.redraw()
+                    }
```
There seems to be some repetition between 61-76 and 35-55. Is there a reason one couldn't use the same code as a subfunction? It looks like the initial instance can be part of the recursive unless I'm missing something.

pattisdr replied:
I separated them because I wanted the first ten to be fetched, and then for the view to render while the rest of the nodes continued to load in the background. The request for the recursive call has the background : true flag set, but I don't want that background:true flag to be used for the first url getting the first ten.

pattisdr replied:
You're right - I have consolidated the recursion.

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+            else {
+                self.sortDateDescending()
+            }
+        };
```
You can probably use **switch** here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+        // For xs screen
+        self.sortFieldGivenDirection = function(){
+            var dropdown = document.getElementById('sortDropDown');
+            var fieldSort = dropdown.options[dropdown.selectedIndex].value;
```
Is there a reason you are not using jquery for this? It's better to use jquery for DOM access.

pattisdr replied:
I must have misunderstood you - you told me not to mix jquery with mithril.

caneruguz replied:
Oh okay I remember now. Yes you are right, it is a bit more nuanced though, the point is to not mix DOM manipulation with mithril in certain ways. In the examples that you have you are still doing DOM manipulation with the vanilla js api, while not using a mithril way. So it's not a huge improvement over jQuery. It's a matter of architecture as well if mithril is not manipulating the element but just producing it then jQuery would be fine but if both of them manipulate it there may be problems. This might need further explanation, find me if you'd like to talk more.

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+                var node = self.nodes()[n];
+                if (self.noTitleMatch(node) && self.noContributorMatch(node) && self.noTagMatch(node)) {
+                    self.nonMatchingNodes().push(node);
+                    self.nodes().splice(n, 1)
```
Very nice workaround to the splicing array problem. :+1:

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+        // Colors node div and changes cursor to pointer on hover
+        self.mouseOver = function (node) {
+            node.style.backgroundColor='#E0EBF3';
+            node.style.cursor = 'pointer'
```
The mouseover classes should be assigned with the `:hover` in CSS. This works better to separate css from javascript files.

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+        };
+
+        self.clearSearch = function () {
+            document.getElementById('searchQuery').value="";
```
Again, you should replace all DOM access and manipulation with jQuery. (http://api.jquery.com/). The above would be for instance ```$('#searchQuery').val('')'```

pattisdr committed:
Add comments to variables in quick search controller.

caneruguz commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,551 @@
+        };
+
+        self.quickSearch = function () {
+            self.filter(document.getElementById('searchQuery').value);
```
For instance here, the filter value should be an m.prop() so that quick search keyup event or onchange event trigger the m.prop change and the filtering function. That's more of a mithril way without using DOM access either through js or jQuery.

pattisdr commented:
Note to self: put class in quotes.

caneruguz commented on website/static/css/quick-project-search-plugin.css:
```diff
@@ -0,0 +1,45 @@
+    border: 1px solid #cccccc;
+}
+
+button {
```
In general it's best to not use very generic identifiers to define rules in CSS. Other aspects of this page will use button. Same with using class names like 'selected' which may be very common.You can either use specific selectors like `.quick-project button`  or namespace individual elements like `.qps-button` I prefer the first method so that you don't add classes to all the elements. The trick is to use a top level unique identifier and use specific classes for things underneath.

caneruguz commented:
CR1 Complete. It looks good to me.  I wouldn't have done some things this way but that's a matter of preference, I think the code logic works fine. 

The writing style is very clean and modular. Very intelligently put together.

pattisdr commented:
What are the big things that you would have done differently?

On Thu, Jan 21, 2016 at 9:23 AM, Caner Uguz <notifications@github.com>
wrote:

> CR1 Complete. It looks good to me. I wouldn't have done some things this
> way but that's a matter of preference, I think the code logic works fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caneruguz/osf.io/pull/17#issuecomment-173583634>.
>

pattisdr committed:
Use switch instead of if/elseif/else.

pattisdr committed:
jshint semicolon fix because javascript requires a million of them.

pattisdr committed:
Assign mouseover classes with hover, separating CSS from JS.

pattisdr committed:
Add top level unique identifier to quick search project class.

pattisdr committed:
Get rid of using js for DOM manipulation of sorting buttons.

pattisdr committed:
Modify search bar so onkeyup event triggers m.prop change of self.filter() instead of using js to access the DOM.

pattisdr committed:
Remove JS access of DOM for sort dropdown.

pattisdr committed:
Rearrange sortDirectionGivenField so not accessing default sort as id.

pattisdr committed:
For xs screen, hide xsDropdown until all nodes are loaded.

pattisdr committed:
Make 'selected/not-selected' CSS classes as classes beneath the broader quick-project class.

pattisdr committed:
Overhaul of quick-search projects - instead of shifting nodes between lists, have an eligible node list with indices referencing the master node list.

pattisdr committed:
Consolidate recursion.

pattisdr committed:
jshint

pattisdr commented:
@caneruguz, I've responded to CR1 - reorganized code so nodes aren't shifted in between lists, instead having one master list of nodes that are referenced with an array of indices.

chadwhitacre commented:
Reviewing ...

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+        }
+
+        function displayNodes() {
+            console.log(ctrl.eligibleNodes().length, ctrl.countDisplayed(), ctrl.sortState(), ctrl.loadingComplete());
```
We can remove this debugging aid, ya?

chadwhitacre commented:
`Foo` doesn't filter the listing as I'd expect, though `Fl` does.

![screen shot 2016-01-26 at 3 19 38 pm](https://cloud.githubusercontent.com/assets/134455/12593691/4f138b8a-c440-11e5-9ff3-92ffa146c911.png)

![screen shot 2016-01-26 at 3 19 46 pm](https://cloud.githubusercontent.com/assets/134455/12593694/51e80b6a-c440-11e5-8300-c5471864337b.png)

pattisdr commented:
Do you have any tags in the 'floo' project called 'foo'? It is searching through title, contributor full names, and tags.

chadwhitacre commented:
@pattisdr No tags, but `Floo` is a component of `Foo`, and has a contributor named `Foo Bar`.

pattisdr commented:
Okay, it is also searching through non-visible contributors.

chadwhitacre commented:
In this case the contributor is visible, but only the last name is displayed in the listing.

pattisdr commented:
And contributor full names.

chadwhitacre commented:
Right. Okay!

chadwhitacre commented:
Weeeeeelllllllllll I guess I'm supposed to review the code, anyway, not so much the UI/visual design. :-)

chadwhitacre commented:
What's supposed to happen on the old dashboard?

![screen shot 2016-01-26 at 3 27 18 pm](https://cloud.githubusercontent.com/assets/134455/12593893/58c9c85a-c441-11e5-97b0-8a59b5889626.png)

pattisdr commented:
I think that's where the new project organizer is going?

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+        self.loadingComplete = m.prop(false); // True when all user nodes are loaded.
+        self.contributorMapping = {}; // Maps node id to list of contributors for searching
+        self.filter = m.prop(); // Search query from user
+        self.fieldSort = m.prop(); // For xs screen, either alpha or date
```
`xs` is a Bootstrap convention meaning "extra small," right?

pattisdr replied:
Yes, that is what I am referring to.

chadwhitacre replied:
M'kay.

chadwhitacre commented:
/me reading up on http://mithril.js.org/getting-started.html ...

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+ * UI and function to quick search projects
+ */
+
+var $ = require('jquery');
```
Don't need this, eh? :)

chadwhitacre commented:
Mithril looks cool. I like the emphasis on functional programming and classic MVC (I've long said that server-side MVC is an oxymoron).

chadwhitacre commented:
> I think that's where the new project organizer is going?

Huh, okay, I'm getting something that looks like it could be a new project organizer. :-)

![screen shot 2016-01-26 at 4 48 15 pm](https://cloud.githubusercontent.com/assets/134455/12596163/aa08fbcc-c44c-11e5-8bef-b508cc4eeab5.png)

chadwhitacre commented on api/users/views.py:
```diff
@@ -288,25 +287,23 @@ class UserNodes(JSONAPIBaseView, generics.ListAPIView, UserMixin, ODMFilterMixin
-        if current_user.is_anonymous():
-            auth = Auth(None)
-        else:
-            auth = Auth(current_user)
```
We're now depending on the [permission checking](https://github.com/CenterForOpenScience/osf.io/blob/d5293f46b384891f114e667013224cd446679f13/api/users/views.py#L42-L44) in `UserMixin.get_user`, ya?

pattisdr replied:
Correct.

pattisdr replied:
Good catch, after looking at this more closely, this query isn't correct.  The main reason I was altering this is that UserNodes get_queryset was returning a list_comprehension, which I didn't want.  I needed some way to filter through the user nodes and sort out which ones were either public, or the user could view without a list_comprehension. I have adjusted the search query to handle permissions properly.

chadwhitacre replied:
`NodeList.get_default_odm_query` [includes](https://github.com/CenterForOpenScience/osf.io/blob/35596266100a1c7848a936d59296bde2fb27b5a1/api/nodes/views.py#L235-L247) a permissions filter that is subtly different from the [new one](https://github.com/CenterForOpenScience/osf.io/blob/35596266100a1c7848a936d59296bde2fb27b5a1/api/users/views.py#L294-L309) in `UserNodes.get_default_odm_query`—another reason to share code between the two (cf. https://github.com/caneruguz/osf.io/pull/17#discussion_r51285543).

chadwhitacre commented on api/users/views.py:
```diff
@@ -288,25 +287,23 @@ class UserNodes(JSONAPIBaseView, generics.ListAPIView, UserMixin, ODMFilterMixin
             Q('is_collection', 'ne', True) &
-            Q('is_deleted', 'ne', True)
+            Q('is_deleted', 'ne', True) &
+            Q('is_registration', 'ne', True)
```
If I'm reading this right, nodes can be of various types, and "folder" and "registration" are two types that we _don't_ want to include in this API. Yes?

chadwhitacre replied:
> "folder" and "registration" are two types that we _don't_ want to include in this API

(Along with "collection" and "deleted".)

chadwhitacre replied:
Is this a query that can be overriden by clients using the querystring? Or is this a baseline query that clients may only _further_ constrain? I ask because we advertise in the docstring that `registration` is filterable, and if the latter then this would seem to be a breaking change.

pattisdr replied:
I was reconciling behavior currently on develop, and Caner's dashboard-redesign-component-3 that I am merging with.  Currently, on develop, the NodeList endpoint /v2/nodes/ only returns nodes that are not deleted, are not folders, and are not registrations.  

/v2/nodes/ query on develop:
![screen shot 2016-01-27 at 9 42 38 am](https://cloud.githubusercontent.com/assets/9755598/12616687/6477610a-c4da-11e5-80b6-c0ad6c427b7c.png)
v2/nodes/ query on Caner's dashboard-redesign-component-3
![screen shot 2016-01-27 at 9 56 51 am](https://cloud.githubusercontent.com/assets/9755598/12617135/4f2775e0-c4dc-11e5-810a-854fd1e8aa76.png)

On Caner's PR, there is instead a restriction on 'collection' instead of 'folder'.  There has been a lot of back and forth as to what should be included in the node list, registrations/no registrations/ but I think the users/me/nodes endpoint has been forgotten.  The users/me/nodes/ endpoint should be returning the same thing as v2/nodes/ except it is further restricted to just the user's projects.  When a user is looking through a list of projects, she shouldn't see deleted projects, or registrations of these projects, or collections(whatever this is), or folders.  So what I have here is doing exactly what /v2/nodes/ is doing except I've taken 'folders' out of the list too.  If folders are in the list, then the 'Dashboard' node comes up, which I certainly don't think is intended to show up in the user nodes list. 

**In summary** My additions to users/me/nodes/
- 'is_folder' excluded in addition to 'is_collection'. We had 'is_folder' before, and without this, the dashboard node shows up.  I also think that 'folders' should be excluded from /v2/nodes/ like they were before unless collections are now folders?
- 'is_registration' excluded.  It is excluded from /v2/nodes/ so it should also be excluded from /v2/users/me/nodes.  There are separate endpoints /v2/registrations/ and v2/users/me/registrations/.

/v2/users/me/nodes/ in my PR:
![screen shot 2016-01-27 at 9 56 11 am](https://cloud.githubusercontent.com/assets/9755598/12617106/333b106c-c4dc-11e5-812d-5e2598ccd501.png)

I know there was a lot of back and forth as to whether registrations should be included in the endpoints, so I think things were changed without bringing all the docs/filtering/other endpoints up to date.  If registrations are excluded from /v2/nodes/ you shouldn't be able to filter on this so it should be removed from filterable fields and the docs.

chadwhitacre replied:
> I think the users/me/nodes endpoint has been forgotten. The users/me/nodes/ endpoint should be returning the same thing as v2/nodes/ except it is further restricted to just the user's projects.

Seems like forgetting users/me/nodes is symptomatic of [WET](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#DRY_vs_WET_solutions) code. Why doesn't `UserNodes` subclass or otherwise reuse the query code from  `NodeList`?

chadwhitacre replied:
>  Is this a query that can be overriden by clients using the querystring? Or is this a baseline query that clients may only _further_ constrain?

Looks like it's a [baseline query](https://github.com/CenterForOpenScience/osf.io/blob/35596266100a1c7848a936d59296bde2fb27b5a1/api/base/filters.py#L290).

> I ask because we advertise in the docstring that `registration` is filterable, and if the latter then this would seem to be a breaking change.

What's our policy on backwards compatibility in the API? Do we care that we're removing functionality (3fecfed195a96aca6d15b830f81e9096e1516ac4) from our documented API within a major version number (v2)?

pattisdr replied:
The way I understand it, we're not claiming backwards compatibility yet - our API is public, but we haven't made a public announcement.  My changes aside, there is an error in the API.  We have separated the nodes and registrations into separate endpoints yet we are still documenting that you can filter on registrations in the nodes endpoint, when none will be returned.

chadwhitacre replied:
Okay.

chadwhitacre commented on api/users/views.py:
```diff
@@ -288,25 +287,23 @@ class UserNodes(JSONAPIBaseView, generics.ListAPIView, UserMixin, ODMFilterMixin
     view_category = 'users'
     view_name = 'user-nodes'
 
+    ordering = ('-date_modified')
```
Does this deserve a documentation update?

chadwhitacre replied:
I don't see any currently documented sort order, so I suppose this isn't a breaking change.

pattisdr replied:
Yes, we are currently filtering v2/nodes/ on -date_modified, so it v2/users/me/nodes/ should probably be filtered the same way.

chadwhitacre replied:
Just realized: this evaluates to a string, when I believe what we want is a tuple. It's a quirk of Python that you need to add a trailing comma to make this a tuple, since otherwise the Python interpreter assumes you meant the parentheses to be used simply for grouping.

```
$ python2
Python 2.7.9 (default, Jan  4 2015, 20:46:16) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> ('foo')
'foo'
>>> ('foo',)
('foo',)
>>>
```

I noticed when comparing with `NodeList` under https://github.com/caneruguz/osf.io/pull/17#issuecomment-180575710. Not sure what the real-world effect would be in this case. Maybe (probably?) our ODM framework is smart enough to handle a string here, in which case we should drop the parens. :)

chadwhitacre commented on webpack.common.config.js:
```diff
@@ -25,6 +25,7 @@ var entry = {
     // JS
     'base-page': staticPath('js/pages/base-page.js'),
     'home-page': staticPath('js/pages/home-page.js'),
+    'landing-page': staticPath('js/pages/landing-page.js'),
```
It's a little weird to have the homepage now be at `/home/`, with `/` called the "landing page" instead, especially since I [gather](https://github.com/caneruguz/osf.io/pull/17#issuecomment-175210892) that the Dashboard isn't going away, either. That's three things that seem to all occupy the same semantic space. How is the ... home page (the quick search implemented here) intended to relate to the dashboard?

chadwhitacre replied:
At this point I'd rather see `home-page.js` still be what shows up at `/`, with a different URL location and JavaScript filename for the quicksearch—though it seems to me that the search functionality on this PR makes more sense as a feature on the new dashboard than as the sole feature on a standalone page. But perhaps @pattisdr @caneruguz you can say more about the direction we're headed? I gather there's a whole Epic [happening](https://openscience.atlassian.net/browse/OSF-5374). ;-)

pattisdr replied:
`/` is the landing_page, which I think is what you see before you've logged in?   

`/home/` is going to be a page with lots of widgets, including my quick search projects which you are reviewing here, also new and noteworthy projects which you've made some comments on, and an activity widget

`/dashboard/` is Caner's project organizer.

This is what's planned for the home page:
![stage 1](https://cloud.githubusercontent.com/assets/9755598/12618710/16666b9c-c4e3-11e5-8c66-c6eb72be245d.png)

pattisdr replied:
Nevermind, a decision was just made to remove /home. https://openscience.atlassian.net/browse/OSF-5605

Once this issue is resolved, I will bring this PR in line with the decision to remove /home.

chadwhitacre replied:
> Nevermind, a decision was just made to remove /home.

@pattisdr Will that be a separate PR? This current PR is the one that _adds_ `/home/` in the first place, right? Should we avoid adding it in the first place, then? :)

pattisdr replied:
No, I didn't add the home route in my PR, I branched off of Caner's dashboard PR and the /home route was already there with Placeholder text.  I just placed the quick search project stuff on that home page.  Multiple people have been developing widgets for the /home page, I'm doing the Quick Search Projects and New and Noteworthy, and Henrique is doing an activity widget.  

Issue https://openscience.atlassian.net/browse/OSF-5605 will be resolved first because it is a blocker to releasing Phase 0 of the dashboard into production (Caner's PR is currently being reviewed: https://github.com/CenterForOpenScience/osf.io/pull/4887/files).  Quick Search projects is part of Phase 1.

chadwhitacre replied:
Cool, thanks for the explanation.

chadwhitacre commented:
##### Punchlist—Done
- [x] "good" test coverage for `{NodeList,UserNodes}.get_default_odm_query`—https://github.com/caneruguz/osf.io/pull/17#issuecomment-180584729
- [x] drop parens or add comma—https://github.com/caneruguz/osf.io/pull/17/files#r52077942 —done in 3fe877c5c91afc6a267fa7df99a3019ce6dbdf87
- [x] refactor `{NodeList,UserNodes}.get_default_odm_query` to share code—https://github.com/caneruguz/osf.io/pull/17#discussion_r51285543 and https://github.com/caneruguz/osf.io/pull/17#discussion_r51289520
  - started in 17adf27882af81d800224a5e9c3135e26d839c74
  - [x] now https://github.com/caneruguz/osf.io/pull/17#issuecomment-180575710 —done in fca592be3befc379a426116cb894df2cc5288930 d3caf6daca98f533be5184482403feef60cc4693 3676b0a99bb8cf602dbe4b8e626fdf48397b1790 and 1917b72cdbf42f56ea110d39da678bc5e96b4f01
- [x] document sort order—https://github.com/caneruguz/osf.io/pull/17#discussion_r50907990 —done in 38065ef07c2239d4be28ca378e3b7c91c49c0bc8
- [x] think through implications of default query change—https://github.com/caneruguz/osf.io/pull/17#discussion_r50907851
- [x] make sure we're not changing authentication behavior—https://github.com/caneruguz/osf.io/pull/17#discussion_r50907480
- [x] remove the jQuery requirement—https://github.com/caneruguz/osf.io/pull/17#discussion-diff-50903861 —done in 53f718837edae4951b827a62cd311f7af27eb87c
- [x] remove a debugging line—https://github.com/caneruguz/osf.io/pull/17#discussion-diff-50893812 —done in 1edb4c4b7f35f4a0ac75cace377d0170c3fd0ce4
- [x] think through confusion between `/`, `/home/`, and `/dashboard/`—https://github.com/caneruguz/osf.io/pull/17#discussion_r50908898
- [x] specify another selector a bit more—https://github.com/caneruguz/osf.io/pull/17#discussion_r50933103 —done in 1edb4c4b7f35f4a0ac75cace377d0170c3fd0ce4

chadwhitacre commented:
It's weird that the left-hand-side magnifying glass and the right-hand-side "X" have the same visual treatment, but one is clickable and the other isn't. Their visual treatment should differ.

![screen shot 2016-01-26 at 8 42 48 pm](https://cloud.githubusercontent.com/assets/134455/12601240/80e3f9ce-c46d-11e5-8006-18855215c475.png)

chadwhitacre commented:
Furthermore, the "X" should visually signal that it's disabled when the search input is empty.

chadwhitacre commented:
@pattisdr Do you want feedback on the visual design? Or should I stick to the code?

cc: @caneruguz @sloria

chadwhitacre commented on website/static/css/quick-project-search-plugin.css:
```diff
@@ -0,0 +1,52 @@
+    }
+}
+
+label {
```
As @caneruguz said [earlier](https://github.com/caneruguz/osf.io/pull/17#discussion_r50405417), a bit more specificity is preferred for selectors, to avoid overselecting.

P.S. This is a place where a [preprocessor](https://openscience.atlassian.net/browse/OSF-5524) can be helpful, because it allows one to [nest selectors](http://sass-lang.com/guide#topic-3) (though this can be [abused](www.sitepoint.com/beware-selector-nesting-sass/)).

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+            }
+            else {
+                return 'not-selected';
+            }
```
This could be a good place for a ternary operator (not asking you to change it on this PR, though).

``` javascript
return self.sortState() === sort ? 'selected' : 'not-selected';
```

chadwhitacre replied:
Especially since this function is so close to the next, using the one-line form would help in scanning the difference between the two (`sortState` vs. `preSelectDirection`).

chadwhitacre replied:
Done in a89d02ed1ba31698b9048fbcea2a1ee7d98a9c30.

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+        self.fieldSort = m.prop(); // For xs screen, either alpha or date
+        self.directionSort = m.prop(); // For xs screen, either Asc or Desc
+
+        // Load first ten nodes
```
The ten count comes from the default page size in the API, ya? I don't see that we're specifically requesting 10 nodes here ...

pattisdr replied:
Yes, so really loads up to ten.

chadwhitacre replied:
k

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+        function sortAlphaAsc() {
+            if (ctrl.loadingComplete()) {
+                return m('button', {id: 'alphaAsc', 'class': ctrl.colorSortButtons('alphaAsc'), onclick: function() {
+                    ctrl.sortBySortState(ctrl.sortState('alphaAsc'));
```
Why not go the last mile and make a `ctrl.sortBy`? So here you could just do ...

``` javascript
ctrl.sortBy('alphaAsc');
```

Idea for next time. :-)

pattisdr replied:
This made more sense when instead of having a master node list, I had multiple lists of nodes that were displayed on the screen, nodes pending in the background, and nodes that didn't match the search query and not eligible to be displayed.  Since I was shifting nodes between lists, I was having to sort them before displaying them...so there was sorting happening even if a user didn't press a sort button. It was sorting by whatever the current sort state was.

But there is still value in tracking what the sort state is, and sorting by that sort state on the xs screen that has different controls than the full screen.

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+                    break;
+                default:
+                    self.sortDateDescending();
+            }
```
@pattisdr I noticed that @caneruguz [steered](https://github.com/caneruguz/osf.io/pull/17#discussion-diff-50315674) you from an `if` tree towards this `switch`. Another nice idiom is to use a hash (object):

``` javascript
{ alphaAsc: self.sortAlphabeticalAscending
, alphaDesc: self.sortAlphabeticalDescending
, dateAsc: self.sortDateAscending
, dateDesc: self.sortDateDescending
, null: self.sortDateDescending
 }[self.sortState()]()
```

Not needed on this PR, though. :)

chadwhitacre commented on website/static/js/quickProjectSearchPlugin.js:
```diff
@@ -0,0 +1,510 @@
+                return (A > B) ? -1 : (A < B) ? 1 : 0;
+            });
+            self.sortState('dateDesc');
+        };
```
Seems like these sort functions could be refactored to duplicate less code. Not a blocker.

chadwhitacre replied:
Addressed in 03b3b0ca70b70040dc6a8597c4a787428223ffd9 (and 35596266100a1c7848a936d59296bde2fb27b5a1).

pattisdr commented:
>  Do you want feedback on the visual design? Or should I stick to the code?

I don't have much experience with html/css/visual design.  I welcome any comments you have about the appearance but it was my understanding that @caneruguz was going to polish the UI after code review.

chadwhitacre commented:
> it was my understanding that @caneruguz was going to polish the UI after code review.

Gotcha. I'll leave that off the [punchlist](https://github.com/caneruguz/osf.io/pull/17#issuecomment-175262395), then. :)

On the code side things look pretty good, and I would second @caneruguz at https://github.com/caneruguz/osf.io/pull/17#issuecomment-173583634. The biggest question I have is about how this new search page is supposed to relate to the new dashboard and how this PR fits into the bigger picture for the dashboard redesign: https://github.com/caneruguz/osf.io/pull/17#discussion_r50908898.

pattisdr committed:
Remove jQuery requirement.  No longer using.

pattisdr committed:
Update documentation to reflect that user nodes are ordered by date_modified.

pattisdr committed:
Remove registration from node filterable fields and update NodeList and UserNodes documentation to reflect this.

pattisdr committed:
Refactor get_default_odm_query for UserNodes view.  If user is anonymous, or logged_in user does not match user in query (users/user_id/nodes/), then user can only see public nodes by this user.

This should return same results as before this PR, except results from get_queryset are not a list comprehension.

pattisdr committed:
Add specificity for 'label' selector and delete print statement.

pattisdr committed:
Use ternary operators for changing class on sort buttons.

pattisdr committed:
Consolidate common line of sorting functions.

pattisdr commented:
@whit537, I have either made requested changes or responded to your comments on the punchlist.  No. 3, making sure we're not changing authentication behavior was a good catch - thanks.

chadwhitacre commented:
@pattisdr Thanks, I expect to be back around here on Friday.

pattisdr committed:
Semicolons.

chadwhitacre commented:
Okay @pattisdr, I believe I'm caught up on your latest commits and comments, with the [punchlist](https://github.com/caneruguz/osf.io/pull/17#issuecomment-175262395) updated accordingly.

chadwhitacre commented:
I'm having some trouble getting a coverage report for the `api` module. I had success with the `website` module over at https://github.com/CenterForOpenScience/osf.io/pull/4799#issuecomment-173943290, but a parallel construction for our case here yields only "Failed to generate report: No data to report." I've confirmed that tracing _is_ on for the `api` files.

```
[osf.io]$ COVERAGE_DEBUG=callers,config,dataio,dataop,pid,plugin,sys,trace py.test --cov=api api_tests/users/views/test_user_nodes_list.py -k test_authorized_in_gets_200 > tracing.api.log 2>&1
```

I'm out of time for today, I'll plan to revisit this on Monday.

chadwhitacre commented:
I've tracked down the coverage problem to [this config stanza](https://github.com/CenterForOpenScience/osf.io/blob/35596266100a1c7848a936d59296bde2fb27b5a1/.coveragerc#L2-L4), which causes the `api` library to get [filtered out](https://github.com/nedbat/coveragepy/blob/coverage-4.0.3/coverage/report.py#L42-L44) of `file_reporters`. Working on a workaround and/or fix ...

chadwhitacre commented:
I'm not finding a way to override the `.coveragerc` settings.

chadwhitacre commented:
Alright, PR to modify `.coveragerc` in https://github.com/CenterForOpenScience/osf.io/pull/4965.

chadwhitacre commented:
Okay!

chadwhitacre commented:
Meanwhile, back at the ranch ...

:cow2:

chadwhitacre commented:
I'd like to see 100% test coverage for the refactored `{NodeList,UserNodes}.get_default_odm_query`; [punchlist](https://github.com/caneruguz/osf.io/pull/17#issuecomment-175262395) tweaked accordingly.

chadwhitacre commented:
I'm looking at (with https://github.com/CenterForOpenScience/osf.io/commit/4b6e9b6660d74800683951d5be2c721ca89e1e48):

```
[osf.io]$ py.test --cov=api.users.views api_tests/users/views/ --cov-report=annotate
```

and:

```
[osf.io]$ py.test --cov=api.nodes.views api_tests/nodes/views/ --cov-report=annotate
```

I'm seeing 100% coverage of the relevant `get_default_odm_query` in both, which is good for refactoring. Of course, line coverage is just the starting point (what about different values? edge cases? complex conditionals?), but I don't think we need to set the plow too deep on this PR.

pattisdr commented:
After asking around, the common code in `get_default_odm_query`  has been abstracted out in https://github.com/CenterForOpenScience/osf.io/pull/4906/files, though it is not part of `develop` yet, or this PR.  So registrations are excluded from both NodeList and UserNodes,  and folders are renamed to collections.  

So I don't need to make any changes to this query, just bring this PR up to date with what is happening on the dashboard.

chadwhitacre commented:
Okay, cool.

pattisdr committed:
Merge remote-tracking branch 'brian/feature/api/remove_registrations_from_nodes_dashboard' into quickSearchProjects

Conflicts:
	api/users/views.py

pattisdr committed:
Ensure registrations are not included in user nodes list.

pattisdr commented:
I merged brian/feature/api/remove_registrations_from_nodes_dashboard which pulled out the common search query between UserNodes and NodeList. I am waiting to bring everything up to date with the dashboard branch when it is no longer fluctuating.

I also added to the UserNodes test confirmation that registrations were not being included in the query.  Are there other tests that you want @whit537?

chadwhitacre commented:
@pattisdr It looks like, even after 17adf27882af81d800224a5e9c3135e26d839c74, we've still got some unnecessary drift between `get_default_odm_query` in [`NodeList`](https://github.com/CenterForOpenScience/osf.io/blob/cdb9b24244d54e896c48558a35e62150e680b33c/api/nodes/views.py#L236-L248) and [`UserList`](https://github.com/CenterForOpenScience/osf.io/blob/cdb9b24244d54e896c48558a35e62150e680b33c/api/users/views.py#L295-L305):
- `NodeList` does not use `default_node_list_query`(?!)
- They duplicate the code for the permission query.

pattisdr commented:
Oh you're right, I thought it was there, but it's on NodeChildrenList.

chadwhitacre commented:
> I also added to the UserNodes test confirmation that registrations were not being included in the query.

Cool, though now that registration inclusion is implemented in a lower-level function (`default_node_list_query`), it's actually better practice to exercise that behavior in the lower-level function rather than the functions that call it. Not a huge deal.

> Are there other tests that you want @whit537?

Not sure ... need to see how the `get_default_odm_query` refactor plays out. :-)

chadwhitacre commented:
[Punchlist](https://github.com/caneruguz/osf.io/pull/17#issuecomment-175262395) updated.

pattisdr committed:
NodeList should use default_node_list_query().

pattisdr committed:
Add comma to make ordering evaluate to a tuple.

pattisdr committed:
Add default_node_permission_query to base_utils.

pattisdr committed:
Use combination of base_query and permissions_query for NodeList get_default_odm_query.

pattisdr committed:
Use combination of base_query and permissions_query for UserNodes get_default_odm_query combined with checking to ensure the contributor is a user.

chadwhitacre commented:
Latest commits look good. Let's see about tests ...

chadwhitacre commented:
I'm seeing coverage of both `get_default_odm_query` methods as well as the `default_*_query` methods. I guess that's enough for me here. :-)

Good work, @pattisdr! Go science! :)

pattisdr commented:
Hooray!

pattisdr committed:
Add line, 'Go to My Projects to organize your work or Search Everything'.

pattisdr committed:
Add link to create projects.

pattisdr committed:
Make text a paragraph instead of h5 header.

pattisdr committed:
Merge branch 'dashboard-redesign-component-3' of https://github.com/caneruguz/osf.io into quickSearchProjects
